compositor_render/src/renderer/scene/constraints.rs

```diff
@@ -0,0 +1,46 @@
+use compositor_common::{
+    error::ConstraintsValidationError,
+    scene::{
+        validation::{constraints::NodeConstraints, inputs::InputsCountConstraint},
+        NodeParams, NodeSpec, SceneSpec,
+    },
+};
+
+pub fn validate_constraints(scene: &SceneSpec) -> Result<(), ConstraintsValidationError> {
+    for (node_id, node_constraints) in scene
+        .nodes
+        .iter()
+        .map(|node| (&node.node_id, node_constraints(node)))
+    {
+        node_constraints.validate(scene, node_id)?
+    }
+
+    Ok(())
+}
```

Comment:
This function will need to take renderer contexts when web renderer and shader constraints will be user / API configurable.
I haven't put it in this PR, to firstly review the general idea for the constraints system.

Reply:
> This function will need to take renderer contexts when web renderer and shader constraints will be user / API configurable.

This function can't take that as param if you define it in common crate.